json/json.go

```diff
@@ -0,0 +1,126 @@
+package json
+
+import (
+	"context"
+	"encoding/json"
+	"net/http"
+	"time"
+
+	"github.com/sourcegraph/zoekt"
+	"github.com/sourcegraph/zoekt/query"
+)
+
+// defaultTimeout is the maximum amount of time a search request should
+// take. This is the same default used by Sourcegraph.
+const defaultTimeout = 20 * time.Second
+
+func JSONServer(searcher zoekt.Searcher) http.Handler {
+	s := jsonSearcher{searcher}
+	mux := http.NewServeMux()
+	mux.HandleFunc("/search", s.jsonSearch)
+	mux.HandleFunc("/list", s.jsonList)
+	return mux
+}
+
+type jsonSearcher struct {
+	Searcher zoekt.Searcher
+}
+
+type jsonSearchArgs struct {
+	Q    string
+	Opts *zoekt.SearchOptions
+}
+
+type jsonSearchReply struct {
+	Result *zoekt.SearchResult
+}
+
+type jsonListArgs struct {
+	Q    string
+	Opts *zoekt.ListOptions
+}
+
+type jsonListReply struct {
+	List *zoekt.RepoList
+}
+
+func (s *jsonSearcher) jsonSearch(w http.ResponseWriter, req *http.Request) {
+	ctx := req.Context()
+	w.Header().Add("Content-Type", "application/json")
+
+	if req.Method != "POST" {
+		jsonError(w, http.StatusMethodNotAllowed, "Only POST is supported")
+		return
+	}
+
+	searchArgs := jsonSearchArgs{}
+	err := json.NewDecoder(req.Body).Decode(&searchArgs)
+	if err != nil {
+		jsonError(w, http.StatusBadRequest, err.Error())
+		return
+	}
+	if searchArgs.Q == "" {
+		jsonError(w, http.StatusBadRequest, "missing query")
+		return
+	}
+	if searchArgs.Opts == nil {
+		searchArgs.Opts = &zoekt.SearchOptions{}
+	}
+
+	query, err := query.Parse(searchArgs.Q)
+	if err != nil {
+		jsonError(w, http.StatusBadRequest, err.Error())
+		return
+	}
+
+	// Set a timeout if the user hasn't specified one.
+	if searchArgs.Opts.MaxWallTime == 0 {
+		var cancel context.CancelFunc
+		ctx, cancel = context.WithTimeout(ctx, defaultTimeout)
+		defer cancel()
+	}
+
+	searchResult, err := s.Searcher.Search(ctx, query, searchArgs.Opts)
```

Comment:
Can you call `searchArgs.Opts.SetDefaults()` before calling Search.

Reply:
This is already being done here, right?

https://github.com/sourcegraph/zoekt/blob/d1964a317f841afecc71b7210fb6ef4c3de26733/eval.go#L144-L147

Based on that, I think the call here is actually redundant.

https://github.com/sourcegraph/zoekt/blob/d1964a317f841afecc71b7210fb6ef4c3de26733/web/server.go#L288-L291